### PR TITLE
Remote queries

### PR DESCRIPTION
**Pull request by gjgd**
- Add a CouchDB db adapter that supports remote querying
- Add a test for the adapter (it is skipped because it requires a prod db url) see instructions for how to obtain a prod cloudant db url 
- Add another sidetree v2 service which uses the Observer VS Resolver
- Implement the
    - Sync will write to the cache the operations contained in the resolved batch files
    - Resolve will only look at the cached operations to resolve a DID (and therefore is fast)
- Separate sidetree, sidetree v2 and sidetree tests into 3 files

This is a light version of sidetree that will work with our needs for beta.
A non exhaustive list of things to implement / transfer from sidetree v1:

- Schema dependency
- Handle errors when batch files are anchor files were not resolved
- Use deterministic stringify for encodeJSON
- Add more tests
- Implement RECOVER and DELETE operations
- Test UPDATE operation
- Test batching
- Add operation validation
- Test with only 1 key in the payload
- Universal resolver
- Element app supports new endpoints

### CODE REVIEW

File: packages/element-api/src/express/routes/sidetree-v2/index.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
+const express = require('express');
+
+const router = express.Router();
+
+router.get('/sync', async (req, res, next) => {
+  try {
+    const sidetree = req.app.get('sidetree-v2');
+    await sidetree.sync();
```
Comment (OR13) [inline packages/element-api/src/express/routes/sidetree-v2/index.js]: This is the kind of call I would expect to be impractical at scale... this is essentially, get me every DID Document that has ever been created...

File: packages/element-api/src/express/routes/sidetree-v2/index.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    res.status(200).send({ ok: true });
+  } catch (e) {
+    next(e);
+  }
+});
+
+router.get('/:did', async (req, res, next) => {
```
Comment (OR13) [inline packages/element-api/src/express/routes/sidetree-v2/index.js]: option to document endpoints describing the benefit of the v2 interface

File: packages/element-lib/src/adapters/database/ElementCouchDBAdapter.js [hunk 3]
```diff
@@ -0,0 +1,79 @@
+const NodeCouchDb = require('node-couchdb');
```
Comment (OR13) [inline packages/element-lib/src/adapters/database/ElementCouchDBAdapter.js]: This Adapter is almost certainly better for backend implementations, but we should note that it is possible to get a couchdb adapter that supports remote queries in the browser: https://github.com/redgeoff/slouch

File: packages/element-lib/src/adapters/database/ElementCouchDBAdapter.spec.js [hunk 4]
```diff
@@ -0,0 +1,113 @@
+const ElementCouchDBAdapter = require('./ElementCouchDBAdapter');
+
+// Needs a prod cloudant URL to run
+describe.skip('ElementCouchDBAdapter', () => {
```
Comment (OR13) [inline packages/element-lib/src/adapters/database/ElementCouchDBAdapter.spec.js]: option we could run couchdb locally in travis and have integration tests for this use case in the api.

File: packages/element-lib/src/sidetree-v2/func.js [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    .update(Buffer.from(toBeSigned))
+    .digest();
+  const privateKeyBuffer = Buffer.from(privateKeyHex, 'hex');
+  const signatureObject = secp256k1.sign(hash, privateKeyBuffer);
+  const signature = base64url.encode(signatureObject.signature);
+  return signature;
+};
+
+
+module.exports = {
```
Comment (OR13) [inline packages/element-lib/src/sidetree-v2/func.js]: I love this minimal interface... once we get some feedback on these changes, I would like to consider moving this stuff to element-core / typescript.

File: packages/element-lib/src/sidetree-v2/index.js [hunk 6]
```diff
@@ -0,0 +1,14 @@
+const { resolve, sync, create } = require('./protocol');
+
+class Sidetree {
+  constructor({ db, blockchain, storage }) {
+    this.blockchain = blockchain;
+    this.storage = storage;
+    this.db = db;
+    this.resolve = resolve(this);
```
Comment (OR13) [inline packages/element-lib/src/sidetree-v2/index.js]: this kind of registration breaks vscode tooling support / making it harder to actually use this library... another reason to switch to typescript.

File: packages/element-lib/src/sidetree-v2/protocol/resolve.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+  const type = operation.decodedOperation.header.operation;
+  switch (type) {
+    case 'create':
+      return {
+        ...operation.decodedOperationPayload,
+        id: `did:elem:${payloadToHash(operation.decodedOperationPayload)}`,
+      };
+    case 'update':
+      return operation.decodedOperationPayload.patch.reduce(jsonpatch.applyReducer, state);
+    case 'recover':
```
Comment (OR13) [inline packages/element-lib/src/sidetree-v2/protocol/resolve.js]: I know that from a compatibility perspective, sidetree-core may not have this functionality implemented, but this makes me a little sad.

File: packages/element-lib/src/sidetree-v2/protocol/resolve.js [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+      return state;
+    case 'delete':
+      return state;
```
Comment (OR13) [inline packages/element-lib/src/sidetree-v2/protocol/resolve.js]: Same for this, I believe these 2 methods are supported in core.

File: packages/element-lib/src/sidetree-v2/protocol/sync.js [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+  syncTransaction,
+  isTransactionValid,
+} = require('../func');
+
+const sync = sidetree => async () => {
+  // Get a set of transactions that have only been processed from cache
+  const transactionsAlreadyProcessed = await sidetree.db.readCollection('transaction');
+  const processedSet = new Set(transactionsAlreadyProcessed.map(t => t.transactionNumber));
+  // Get all transactions from the smart contract
+  const transactions = await sidetree.blockchain.getTransactions(
```
Comment (OR13) [inline packages/element-lib/src/sidetree-v2/protocol/sync.js]: This looks very expensive. I think ideally the node should always store transactions, there is no reason we should be getting 100% of all transactions ever created from ethereum every-time sync is called.

File: packages/element-lib/src/sidetree-v2/test-utils.js [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    mnemonic: 'hazard pride garment scout search divide solution argue wait avoid title cave',
+    hdPath: "m/44'/60'/0'/0/0",
+    providerUrl: 'http://localhost:8545',
+    anchorContractAddress: '0x1DABA81D326Ae274d5b18111440a05cD9581b305',
+  });
+
+  return new element.SidetreeV2({ db, storage, blockchain });
+};
+
+const getCreatePayload = async () => {
```
Comment (OR13) [inline packages/element-lib/src/sidetree-v2/test-utils.js]: This method can use the `op` shorthand, and should...

**Comment by OR13** [comment]
Per our discussion, lets refactor this to achieve the following objectives:

- 1 set of endpoints for sidetree
- full node sync + just in time resolution (blocking)

Subsequent PRs should focus on typescript support and sidetree-core interop.

File: packages/element-lib/src/adapters/database/ElementCouchDBAdapter.spec.js (hunk 4 above)
Comment (gjgd) [reply packages/element-lib/src/adapters/database/ElementCouchDBAdapter.spec.js]: Created an issue for that: https://github.com/decentralized-identity/element/issues/109

File: packages/element-lib/src/sidetree-v2/protocol/sync.js (hunk 9 above)
Comment (gjgd) [reply packages/element-lib/src/sidetree-v2/protocol/sync.js]: Yes! Let's merge this and I will do that in a separate PR

**Pull request merged**